pkg/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,282 @@
+	} else if newApp.Spec.Mode == v1beta1.ClusterMode {
+		return v.syncPodGroupInClusterMode(newApp)
+	}
+	return newApp, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClientMode(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We only care about the executor pods in client mode
+	newApp := app.DeepCopy()
+	if _, ok := newApp.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		//Only executor resource will be considered.
+		if err := v.syncPodGroup(newApp, 1, getExecutorRequestResource(app)); err == nil {
+			newApp.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(newApp)
+		} else {
+			return nil, err
+		}
+	}
+	return newApp, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClusterMode(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We need both mark Driver and Executor when submitting
+	//NOTE: In cluster mode, the initial size of PodGroup is set to 1 in order to schedule driver pod first.
+	if _, ok := app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		//Both driver and executor resource will be considered.
+		totalResource := sumResourceList([]corev1.ResourceList{getExecutorRequestResource(app), getDriverRequestResource(app)})
+		if err := v.syncPodGroup(app, 1, totalResource); err == nil {
+			app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+			app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		} else {
+			return nil, err
+		}
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) getAppPodGroupName(app *v1beta1.SparkApplication) string {
+	return fmt.Sprintf("spark-%s-pg", app.Name)
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroup(app *v1beta1.SparkApplication, size int32, minResource corev1.ResourceList) error {
+	var err error
+	podGroupName := v.getAppPodGroupName(app)
+	if pg, err := v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Get(podGroupName, v1.GetOptions{}); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		podGroup := v1alpha2.PodGroup{
+			ObjectMeta: v1.ObjectMeta{
+				Namespace: app.Namespace,
+				Name:      podGroupName,
+				OwnerReferences: []v1.OwnerReference{
+					*v1.NewControllerRef(app, v1beta1.SchemeGroupVersion.WithKind("SparkApplication")),
+				},
+			},
+			Spec: v1alpha2.PodGroupSpec{
+				MinMember:    size,
+				MinResources: &minResource,
+			},
+		}
+
+		_, err = v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Create(&podGroup)
+	} else {
+		if pg.Spec.MinMember != size {
+			pg.Spec.MinMember = size
+			_, err = v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Update(pg)
+		}
+	}
+	return fmt.Errorf("failed to sync PodGroup with error: %s. Abandon schedule pods via volcano", err)
+}
+
+func New(config *rest.Config, webhookEnabled bool) (schedulerinterface.BatchScheduler, error) {
+
+	if !webhookEnabled {
```

Comment:
Looks like we don't need this check and `webhookEnabled` any more.

Reply:
How about the scheduler name for the pods?

Reply:
Why do we need to consider the scheduler name here? I meant the check on `webhookEnabled` is no longer needed since we don't use the webhook to apply the batch scheduling mutations, so this check doesn't make sense any more.

Reply:
Volcano only focus on the pod that configured with schedulername `volcano`. and webhook is responsible for passing that value from SparkApplication to the driver/executor pods.

Reply:
Adding annotations is done without the webhook. Spark on k8s has Spark config options for adding driver/executor annotations.

Reply:
Got it. This check makes sense. However, I think the check applies to all batch schedulers and should be done in `main` to make sure when `batch-scheduler-name` is not empty, `enable-webhook` is set to `true`.